File: axon-csys/c-src/probe/probe.c
/*
 * v1.19.0 — axon-csys build-infrastructure probe kernel.
 *
 * The probe is the canary that proves the entire C build pipeline works
 * end-to-end (cc-rs orchestration → C23 compilation → static link → FFI
 * invocation from Rust). Subsequent steps (25.c onward) reuse the same
 * plumbing for real kernels. The probe reports compile-time C-standard
 * realisation + feature availability so tests + downstream kernels can
 * degrade gracefully on toolchains that pre-date individual C23 features.
 *
 * Layout discipline:
 *   • PUBLIC ABI is everything prefixed `axon_csys_probe_*`. Order of
 *     functions in this file is the order in which Rust's extern block
 *     declares them (src/probe.rs).
 *   • All public functions are [[nodiscard]] — there is no fire-and-forget
 *     entry on this surface.
 *   • Feature-detection macros use the standards-mandated
 *     `__has_c_attribute` / `__has_embed` / `__BITINT_MAXWIDTH__` probes;
 *     anything not detectable through standard macros is gated by a
 *     compiler-vendor #ifdef (only for the goto-labels-as-values extension).
 */

#include <stdint.h>
#include <stddef.h>

/* axon-csys requires at least C11 for _Alignas / _Alignof. C23 is the
 * preferred floor (D2). We accept C17 here so that stragglers in CI matrix
 * (e.g. older Apple clang on macOS-12 runners) still build the probe — but
 * only the probe; future kernels (25.c onward) tighten the floor to C23. */
#if !defined(__STDC_VERSION__) || __STDC_VERSION__ < 201112L
#  error "axon-csys requires at least C11 (preferably C23). Upgrade your toolchain."
#endif

/* ---------------------------------------------------------------------------
 * Feature detection — bitmask reported back to Rust via probe_features().
 * Bit positions MUST stay synchronised with AxonCsysFeatures in src/probe.rs;
 * tests in tests/probe.rs assert the round-trip.
 * ------------------------------------------------------------------------- */

#define AXON_CSYS_FEATURE_C23           (1u << 0)
#define AXON_CSYS_FEATURE_EMBED         (1u << 1)
#define AXON_CSYS_FEATURE_BITINT        (1u << 2)
#define AXON_CSYS_FEATURE_UNSEQUENCED   (1u << 3)
#define AXON_CSYS_FEATURE_NULLPTR       (1u << 4)
#define AXON_CSYS_FEATURE_ALIGNAS_64    (1u << 5)
#define AXON_CSYS_FEATURE_COMPUTED_GOTO (1u << 6)

#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 202311L
#  define AXON_CSYS_HAS_C23 1
#else
#  define AXON_CSYS_HAS_C23 0
#endif

/* C23 section 6.10.3.5 mandates __STDC_EMBED_NOT_FOUND__ / __STDC_EMBED_FOUND__ /
 * __STDC_EMBED_EMPTY__ predefined macros once the implementation supports
 * #embed. Their presence is the standard-blessed way to detect support. */
#if defined(__STDC_EMBED_NOT_FOUND__)
#  define AXON_CSYS_HAS_EMBED 1
#else
#  define AXON_CSYS_HAS_EMBED 0
#endif

/* _BitInt(N) availability is signalled by __BITINT_MAXWIDTH__; clang ≥16
 * + gcc ≥14 define it. MSVC does not yet ship _BitInt. */
#if defined(__BITINT_MAXWIDTH__) && __BITINT_MAXWIDTH__ >= 8
#  define AXON_CSYS_HAS_BITINT 1
#else
#  define AXON_CSYS_HAS_BITINT 0
#endif

#if defined(__has_c_attribute)
#  if __has_c_attribute(unsequenced)
#    define AXON_CSYS_HAS_UNSEQUENCED 1
#  else
#    define AXON_CSYS_HAS_UNSEQUENCED 0
#  endif
#else
#  define AXON_CSYS_HAS_UNSEQUENCED 0
#endif

/* `nullptr` constant: C23 keyword. */
#if AXON_CSYS_HAS_C23
#  define AXON_CSYS_HAS_NULLPTR 1
#else
#  define AXON_CSYS_HAS_NULLPTR 0
#endif

/* Computed gotos (labels-as-values): GCC + clang extension; not standard
 * C; MSVC has no equivalent. Used by the FSM dispatcher in 25.e. */
#if defined(__GNUC__) || defined(__clang__)
#  define AXON_CSYS_HAS_COMPUTED_GOTO 1
#else
#  define AXON_CSYS_HAS_COMPUTED_GOTO 0
#endif

/* Cache-line alignment via _Alignas (C11 syntax — works on C11/C17/C23
 * everywhere we care about). C23's `alignas` keyword is just a spelling
 * change. */
#define AXON_CSYS_HAS_ALIGNAS_64 1

/* ---------------------------------------------------------------------------
 * [[nodiscard]] portable guard.
 * `__has_c_attribute(nodiscard)` is C23-mandated; in pre-C23 we compile
 * the attribute away.
 *
 * Pre-C23 GCC (≤ ~10 on aarch64-linux-musl in cross 0.2.5) does NOT define
 * the `__has_c_attribute` builtin and does NOT short-circuit it inside a
 * `#if defined(X) && X(y)` expression — it eagerly expands the second
 * conjunct and emits "missing binary operator before token (". Nested
 * `#ifdef` / `#if` guards avoid the eager-expansion path on those
 * pre-C23 toolchains.
 * ------------------------------------------------------------------------- */

#ifdef __has_c_attribute
#  if __has_c_attribute(nodiscard)
#    define AXON_CSYS_NODISCARD [[nodiscard]]
#  endif
#endif
#ifndef AXON_CSYS_NODISCARD
#  define AXON_CSYS_NODISCARD
#endif

/* ---------------------------------------------------------------------------
 * Cache-line aligned canary struct.
 * Tests assert _Alignof(struct ...) == 64 on every supported platform.
 * ------------------------------------------------------------------------- */

struct axon_csys_cacheline_canary {
    _Alignas(64) uint64_t marker;
    uint8_t  padding[56];
};

/* ---------------------------------------------------------------------------
 * Public ABI — keep stable. New fields go at the END of structs only.
 * The order of declarations here mirrors src/probe.rs's extern block.
 * ------------------------------------------------------------------------- */

/* Returns the axon-csys probe ABI version: (major << 16) | (minor << 8) | patch.
 * Bumped when the FFI surface changes shape, NOT for kernel-internal changes. */
AXON_CSYS_NODISCARD
uint32_t axon_csys_probe_version(void) {
    /* 25.b ships ABI version 0.1.0 — probe kernel only. */
    return (0u << 16) | (1u << 8) | 0u;
}

/* Returns the realised __STDC_VERSION__ at the C kernel's compile site.
 * Tests use this to verify the chosen `-std=` flag actually took effect.
 * Expected values:
 *   201112L = C11
 *   201710L = C17
 *   202000L = C2x (pre-ratification spelling)
 *   202311L = C23 (ratified)
 */
AXON_CSYS_NODISCARD
uint32_t axon_csys_probe_c_standard(void) {
    return (uint32_t)__STDC_VERSION__;
}

/* Returns a bitmask of available C23 features. Tests use this to skip
 * branches that the local toolchain cannot compile. */
AXON_CSYS_NODISCARD
uint32_t axon_csys_probe_features(void) {
    uint32_t mask = 0u;
    if (AXON_CSYS_HAS_C23)           mask |= AXON_CSYS_FEATURE_C23;
    if (AXON_CSYS_HAS_EMBED)         mask |= AXON_CSYS_FEATURE_EMBED;
    if (AXON_CSYS_HAS_BITINT)        mask |= AXON_CSYS_FEATURE_BITINT;
    if (AXON_CSYS_HAS_UNSEQUENCED)   mask |= AXON_CSYS_FEATURE_UNSEQUENCED;
    if (AXON_CSYS_HAS_NULLPTR)       mask |= AXON_CSYS_FEATURE_NULLPTR;
    if (AXON_CSYS_HAS_ALIGNAS_64)    mask |= AXON_CSYS_FEATURE_ALIGNAS_64;
    if (AXON_CSYS_HAS_COMPUTED_GOTO) mask |= AXON_CSYS_FEATURE_COMPUTED_GOTO;
    return mask;
}

/* Sanity FFI call — pure integer arithmetic. Used by tests to confirm
 * that calling C from Rust works at the most basic level. */
AXON_CSYS_NODISCARD
int32_t axon_csys_probe_add(int32_t a, int32_t b) {
    return a + b;
}

/* Returns _Alignof(struct axon_csys_cacheline_canary). MUST be 64 on every
 * platform we ship. Caught at runtime by the Rust drift gate. */
AXON_CSYS_NODISCARD
size_t axon_csys_probe_cacheline_alignment(void) {
    return _Alignof(struct axon_csys_cacheline_canary);
}

/* Round-trips a 64-bit marker through a cache-line aligned struct.
 * Confirms (a) struct lays out correctly across the FFI boundary,
 * (b) auto-storage allocation honours the alignment requirement,
 * (c) reads and writes don't get reordered into adjacent padding. */
AXON_CSYS_NODISCARD
uint64_t axon_csys_probe_cacheline_marker(uint64_t marker) {
    struct axon_csys_cacheline_canary canary;
    canary.marker = marker;
    for (size_t i = 0; i < sizeof canary.padding; ++i) {
        canary.padding[i] = 0u;
    }
    return canary.marker;
}

/* Returns sizeof(struct axon_csys_cacheline_canary). MUST be a multiple of
 * 64 (same reasoning as alignment). Tests assert it. */
AXON_CSYS_NODISCARD
size_t axon_csys_probe_cacheline_size(void) {
    return sizeof(struct axon_csys_cacheline_canary);
}

File: axon-csys/tests/probe.rs
//! v1.19.0 — End-to-end build-infrastructure probe tests.
//!
//! These tests exercise the entire cc-rs → C23 → static link → FFI
//! pipeline. If they pass on Linux/macOS/Windows × clang/gcc/msvc, the
//! foundation for 25.c onward is sound.

use axon_csys::{
    probe_add, probe_c_standard, probe_cacheline_alignment, probe_cacheline_marker,
    probe_cacheline_size, probe_features, probe_version, AxonCsysFeatures, AxonCsysVersion,
};

// ────────────────────────────────────────────────────────────────────────
// Basic FFI sanity
// ────────────────────────────────────────────────────────────────────────

#[test]
fn probe_version_is_zero_one_zero() {
    let v = probe_version();
    assert_eq!(
        v,
        AxonCsysVersion {
            major: 0,
            minor: 1,
            patch: 0
        }
    );
    assert_eq!(v.raw(), 0x000100, "wire encoding must match probe.c");
}

#[test]
fn probe_add_arithmetic_is_correct() {
    assert_eq!(probe_add(0, 0), 0);
    assert_eq!(probe_add(1, 2), 3);
    assert_eq!(probe_add(-5, 5), 0);
    assert_eq!(probe_add(i32::MAX - 1, 1), i32::MAX);
    // Commutativity is a free property of integer addition; assert it
    // crosses the FFI boundary intact.
    assert_eq!(probe_add(7, 11), probe_add(11, 7));
}

// ────────────────────────────────────────────────────────────────────────
// C standard realisation — proves the -std= flag chain in build.rs
// actually took effect.
// ────────────────────────────────────────────────────────────────────────

#[test]
fn probe_c_standard_meets_c11_floor() {
    let std = probe_c_standard();
    // axon-csys requires at least C11 (probe.c #error guards this).
    assert!(
        std >= 201112,
        "C standard {std} below C11 floor — toolchain too old for axon-csys",
    );
}

#[test]
fn probe_c_standard_is_a_known_value() {
    let std = probe_c_standard();
    // Sanity: should be one of the published __STDC_VERSION__ values.
    // MSVC quirk: `/std:clatest` reports 202312L instead of the ratified
    // 202311L — observed empirically on MSVC 19.41+. The implementation
    // honours C23 semantics so we accept it as the C23 floor.
    let known = [
        201112u32, // C11
        201710,    // C17
        202000,    // C2x (pre-ratification spelling)
        202311,    // C23 (ratified — clang ≥18, gcc ≥14)
        202312,    // C23 (MSVC `/std:clatest` quirk — see comment above)
    ];
    assert!(
        known.contains(&std),
        "unexpected __STDC_VERSION__ value {std}; expected one of {known:?} (C11/C17/C2x/C23)",
    );
}

// ────────────────────────────────────────────────────────────────────────
// Feature bitmask round-trip + cross-checks against the C standard.
// ────────────────────────────────────────────────────────────────────────

#[test]
fn probe_features_reports_at_least_alignas() {
    // alignas_64 is the one feature we mandate at the C11 floor — every
    // supported toolchain must report it.
    let features = probe_features();
    assert!(
        features.contains(AxonCsysFeatures::ALIGNAS_64),
        "every supported toolchain MUST honour _Alignas(64); got features={:#x}",
        features.raw(),
    );
}

#[test]
fn probe_features_c23_implies_nullptr() {
    // Logical invariant: C23 ⇒ nullptr. If the C kernel reports C23 but
    // not nullptr, the probe macros are inconsistent.
    let features = probe_features();
    if features.contains(AxonCsysFeatures::C23) {
        assert!(
            features.contains(AxonCsysFeatures::NULLPTR),
            "C23 must imply nullptr availability; got features={:#x}",
            features.raw(),
        );
    }
}

#[test]
fn probe_features_c23_flag_matches_stdc_version() {
    // Cross-check: AXON_CSYS_FEATURE_C23 bit is set iff __STDC_VERSION__
    // is ≥202311L.
    let features = probe_features();
    let std = probe_c_standard();
    assert_eq!(
        features.contains(AxonCsysFeatures::C23),
        std >= 202311,
        "C23 feature bit + __STDC_VERSION__ disagree (features={:#x}, std={})",
        features.raw(),
        std,
    );
}

#[test]
fn probe_features_round_trip_through_raw() {
    let features = probe_features();
    let raw = features.raw();
    let reconstructed = AxonCsysFeatures::from_raw(raw);
    assert_eq!(features, reconstructed);
}

// ────────────────────────────────────────────────────────────────────────
// Cache-line layout — proves _Alignas(64) survives the FFI boundary
// and matches Rust's layout expectations.
// ────────────────────────────────────────────────────────────────────────

#[test]
fn probe_cacheline_alignment_is_sixty_four() {
    assert_eq!(
        probe_cacheline_alignment(),
        64,
        "_Alignof(struct axon_csys_cacheline_canary) must be 64",
    );
}

#[test]
fn probe_cacheline_size_is_multiple_of_alignment() {
    let size = probe_cacheline_size();
    let align = probe_cacheline_alignment();
    assert_eq!(
        size % align,
        0,
        "struct size {size} must be a multiple of alignment {align}",
    );
    // The canary is u64 (8) + u8[56] = 64 bytes total — exactly one cache line.
    assert_eq!(size, 64, "canary struct should be exactly one cache line");
}

#[test]
fn probe_cacheline_marker_round_trips() {
    for marker in [
        0u64,
        1,
        u64::MAX,
        0xDEAD_BEEF_CAFE_BABE,
        0x0123_4567_89AB_CDEF,
        u64::MAX / 2,
    ] {
        assert_eq!(
            probe_cacheline_marker(marker),
            marker,
            "marker {marker:#x} did not survive struct round-trip",
        );
    }
}

// ────────────────────────────────────────────────────────────────────────
// Concurrency — confirms the probe is reentrant + thread-safe (it should
// be: all probe functions are pure / stateless).
// ────────────────────────────────────────────────────────────────────────

#[test]
fn probe_is_thread_safe() {
    use std::sync::atomic::{AtomicUsize, Ordering};
    use std::sync::Arc;
    use std::thread;

    let mismatches = Arc::new(AtomicUsize::new(0));
    let mut handles = Vec::new();

    for thread_idx in 0i32..8 {
        let mismatches = Arc::clone(&mismatches);
        handles.push(thread::spawn(move || {
            for i in 0i32..1_000 {
                let a = thread_idx * 1_000 + i;
                let b = i % 17;
                if probe_add(a, b) != a + b {
                    mismatches.fetch_add(1, Ordering::Relaxed);
                }
                let marker = ((thread_idx as u64) << 32) | (i as u64);
                if probe_cacheline_marker(marker) != marker {
                    mismatches.fetch_add(1, Ordering::Relaxed);
                }
            }
        }));
    }

    for h in handles {
        h.join().expect("worker thread panicked");
    }

    assert_eq!(
        mismatches.load(Ordering::Relaxed),
        0,
        "probe must be deterministic + thread-safe under concurrent load",
    );
}
